File: src/Offering/OfferingTest/OfferTest.cs
using Offering.Models;

namespace OfferingTest;

public class OfferTest
{
    [Fact]
    public void Add_Discount_To_Offer()
    {
        var price = Price.CreateFromGross(119, 19, "EUR");
        var product = Product.Create(Guid.NewGuid(), "test product", "test description");
        var offer = SingleOffer.Create("test offer", price, DateTime.Now, DateTime.MaxValue, product);
        
        var discount = Discount.Create(10, DateTime.Now, DateTime.MaxValue);
        offer.ApplyDiscount(discount);
        
        Assert.NotNull(offer.Discount);
        Assert.Equal(107.1m, offer.Price.GrossPrice);
        Assert.Equal(90m, offer.Price.NetPrice);
    }

    [Fact]
    public void Remove_Discount_From_Offer()
    {
        var price = Price.CreateFromGross(119, 19, "EUR");
        var product = Product.Create(Guid.NewGuid(), "test product", "test description");
        var offer = SingleOffer.Create("test offer", price, DateTime.Now, DateTime.MaxValue, product);
        var discount = Discount.Create(10, DateTime.Now, DateTime.MaxValue);
        offer.ApplyDiscount(discount);
        
        offer.RemoveDiscount();
        
        Assert.Null(offer.Discount);
        Assert.Equal(119, offer.Price.GrossPrice);
        Assert.Equal(100, offer.Price.NetPrice);
    }

    [Fact]
    public void Single_Offer_Is_Instantiated_Correctly()
    {
        var price = Price.CreateFromGross(119, 19, "EUR");
        var product = Product.Create(Guid.NewGuid(), "test product", "test description");
        var dateStart = DateTime.Now;
        var dateEnd = DateTime.MaxValue;
        
        var offer = SingleOffer.Create("test offer", price, dateStart, dateEnd, product);
        
        Assert.Equal("test offer", offer.Name);
        Assert.Equal(price, offer.Price);
        Assert.Equal(dateStart, offer.StartDate);
        Assert.Equal(dateEnd, offer.EndDate);
        Assert.Equal(product, offer.Product);
        Assert.Null(offer.Discount);
    }

    [Fact]
    public void Package_Offer_Is_Instantiated_Correctly()
    {
        var price = Price.CreateFromGross(119, 19, "EUR");
        var product1 = Product.Create(Guid.NewGuid(), "test product", "test description");
        var product2 = Product.Create(Guid.NewGuid(), "test product2", "test description");
        var products = new List<Product> {product1, product2};
        var dateStart = DateTime.Now;
        var dateEnd = DateTime.MaxValue;
        var offer = PackageOffer.Create("test offer", price, dateStart, dateEnd, products);
        
        Assert.Equal("test offer", offer.Name);
        Assert.Equal(price, offer.Price);
        Assert.Equal(dateStart, offer.StartDate);
        Assert.Equal(dateEnd, offer.EndDate);
        Assert.Equal(products, offer.Products);
        Assert.Null(offer.Discount);
    }

    [Fact]
    public void Cannot_Apply_2_Discounts_To_Offer()
    {
        var price = Price.CreateFromGross(119, 19, "EUR");
        var product = Product.Create(Guid.NewGuid(), "test product", "test description");
        var offer = SingleOffer.Create("test offer", price, DateTime.Now, DateTime.MaxValue, product);
        var discount1 = Discount.Create(10, DateTime.Now, DateTime.MaxValue);
        offer.ApplyDiscount(discount1);
        
        var discount2 = Discount.Create(20, DateTime.Now, DateTime.MaxValue);
        
        Assert.Throws<InvalidOperationException>(() => offer.ApplyDiscount(discount2));
    }
    
}

File: src/OfferSB/src/test/java/com/example/offersb/test/UnitTests/OfferTest.java
package com.example.offersb.test.UnitTests;

import com.example.offersb.models.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.UUID;

public class OfferTest {

    @Test
    public void testAdd_Discount_To_Offer() {
        var price = Price.CreateFromGross(119, 19);
        var product = Product.create(UUID.randomUUID(), "Test Product", "Test Product Description");
        var localization = Localization.create("DE", "Germany", "Deutschland", "EUR");
        var offer = SingleOffer.create("Test Offer", price, LocalDateTime.now(), LocalDateTime.MAX, product, localization);

        var discount = Discount.create(10, LocalDateTime.now(), LocalDateTime.MAX);
        offer.applyDiscount(discount);

        assert offer.getDiscount() != null;
        assert offer.getPrice().getGrossPrice() == 107.1;
        assert offer.getPrice().getNetPrice() == 90;
    }

    @Test
    public void testRemove_Discount_From_Offer() {
        var price = Price.CreateFromGross(119, 19);
        var product = Product.create(UUID.randomUUID(), "Test Product", "Test Product Description");
        var localization = Localization.create("DE", "Germany", "Deutschland", "EUR");
        var offer = SingleOffer.create("Test Offer", price, LocalDateTime.now(), LocalDateTime.MAX, product, localization);
        var discount = Discount.create(10, LocalDateTime.now(), LocalDateTime.MAX);
        offer.applyDiscount(discount);

        offer.removeDiscount();

        assert offer.getDiscount() == null;
        assert offer.getPrice().getGrossPrice() == 119;
        assert offer.getPrice().getNetPrice() == 100;
    }

    @Test
    public void testPackageOffer_Is_Instantiated_Correctly() {
        var price = Price.CreateFromGross(119, 19);
        var product1 = Product.create(UUID.randomUUID(), "test product", "test description");
        var product2 = Product.create(UUID.randomUUID(), "test product2", "test description");
        var localization = Localization.create("DE", "Germany", "Deutschland", "EUR");
        var products = new ArrayList<Product>(Arrays.asList(product1, product2));
        var dateStart = LocalDateTime.now();
        var dateEnd = LocalDateTime.MAX;
        var offer = PackageOffer.create("Test Offer", price, dateStart, dateEnd, products, localization);

        assert offer.getId() != null;
        assert offer.getName() == "Test Offer";
        assert offer.getPrice() == price;
        assert offer.getStartDate() == dateStart;
        assert offer.getEndDate() == dateEnd;
        assert offer.getProducts() == products;
    }

    @Test
    public void testCannot_Apply_2_Discounts_To_Offer() {
        var price = Price.CreateFromGross(119, 19);
        var product = Product.create(UUID.randomUUID(), "Test Product", "Test Product Description");
        var localization = Localization.create("DE", "Germany", "Deutschland", "EUR");
        var offer = SingleOffer.create("Test Offer", price, LocalDateTime.now(), LocalDateTime.MAX, product, localization);
        var discount1 = Discount.create(10, LocalDateTime.now(), LocalDateTime.MAX);
        offer.applyDiscount(discount1);

        var discount2 = Discount.create(10, LocalDateTime.now(), LocalDateTime.MAX);

        Assertions.assertThrows(IllegalStateException.class, () -> offer.applyDiscount(discount2));
    }
}
